PrivacyPolicy-TOS: add PrivacyPolicyBanner component

retrofox opened the pull request:
~This PR depends on other ones. Take a look at #18980 before to merger this one.~
----------------------------------------------------------------------------------

This patch implements the privacy policy banner into the application. It's the final part of #18980.
The idea is showing a notification banner in the ~plans pages~ whole application when the privacy policy changes to let to the customers know about this.

![image](https://user-images.githubusercontent.com/77539/32073979-583315b6-ba6e-11e7-8fd8-92bf1017c2cb.png)

![privacy-policy-dialog-02](https://user-images.githubusercontent.com/77539/32218879-a3d16f6a-be2c-11e7-8e91-df811c06ad88.gif)

The logic for this starts with the data gotten from the privacy-policy endpoint. The body has, among all data, the notification period and the version of the policy. The sync data process is done in #18975 and #18972.

If the current date is between the notification period and the read-privacy-policy flag is not true then the banner shows. This flag is built based on the privacy-policy entity, type, and version. It's obtained from the response body of the endpoint as well.

Once the user opens the dialog with the privacy policy data the flag sets to true into the user preferences (server-side). At this point, we consider that the user already read the privacy policy.

### Testing

**Note**
_Since the privacy policy has defined a notification period and maybe when you test this feature the current date is out of this period I've added a `privacy-policy/test` flag in order to force showing the banner beyond of the read/accept the status of the user. Restart  calypso enabling this flag:_

```cli
> ENABLE_FEATURES=privacy-policy/test npm start
```

#### Requestion the privacy-policy data

The first thing that you could try is detecting is the request of the privacy-policy endpoint. Just open the dev tools, focus the network tab, and filter the requests to get the privacy-policy one. ~Keep in mind the banner is allowed on `my plan` and `plans` pages:~

After a hard refresh, you should see something like the following:

![image](https://user-images.githubusercontent.com/77539/32219927-1c78c834-be30-11e7-80b7-f8aca3b3e328.png)

![image](https://user-images.githubusercontent.com/77539/32219938-28294dca-be30-11e7-9ced-74f5d238e32f.png)

#### Syncing privacy-policy request

Using the Redux dev extension you could check how the data get from the endpoint is propagated into the state-tree.

![image](https://user-images.githubusercontent.com/77539/32220048-84c23cae-be30-11e7-85b6-2d6767f7f709.png)

<img src="https://user-images.githubusercontent.com/77539/32220081-a6706f2e-be30-11e7-897a-a028fc196b23.png" witdh="600px" />

#### Reading the privacy policy - Setting user preference

When the user closes the dialog the flag according to the privacy policy version sets true. You could detect this request in a similar way that above:

![image](https://user-images.githubusercontent.com/77539/32220252-5a4b802e-be31-11e7-924e-e57dc57dede9.png)

<img src="https://user-images.githubusercontent.com/77539/32220288-7e99f5aa-be31-11e7-9172-3a1da7317e8e.png" witdh="600px" />

_note: I've been testing with many different versions. You should see only one version into the `privacy_policy` field.

The action can be detected as well:

![privacy-policy-redux](https://user-images.githubusercontent.com/77539/32220505-3fbf5d88-be32-11e7-8b5f-5a6a95e51564.gif)

lamosty commented:
What about introducing a feature flag? We could merge smaller parts of the codebase then while not having to worry that a work-in-progress will be shown to customers.

retrofox commented:
> What about introducing a feature flag?

Definitely we are going to add a flag for this.

gwwar commented:
Note that in a dev env, dispatch is available on the window object. So we can trigger dispatches manually. This can make it handy to test if network requests are properly triggered. 

In Console
```
dispatch( { type: 'PRIVACY_POLICY_REQUEST' } );
```

And we can verify the current shape by inspecting the tree:
```
state.privacyPolicy.entities
```

gwwar commented on client/components/data/query-privacy-policy/index.jsx:
```diff
@@ -0,0 +1,30 @@
+ * External dependencies
+ *
+ * @format
+ */
```
/** @format */
/**
 * External dependencies
 */
import PropTypes from 'prop-types';

gwwar commented on client/blocks/privacy-policy-banner/index.jsx:
```diff
@@ -0,0 +1,148 @@
+		}
+
+		// check if the user has already accepted/read the privacy policy.
+		if ( isPolicyAlreadyAccepted === true && ! config.isEnabled( 'privacy-policy/test' ) ) {
```
If the feature flag isn't enabled, wouldn't we return null instead?

retrofox replied:
We have two testing flags. The first one is the feature one named `privacy-policy`. It's used above.
But in this line, we use `privacy-policy/test` to force to show the banner beyond if the user has already accepted/seen the dialog?
You could modify the endpoint version but it isn't practical to do.

retrofox replied:
btw `config.isEnabled( 'privacy-policy/test' )` will be removed once it's landed.

gwwar commented on client/blocks/privacy-policy-banner/index.jsx:
```diff
@@ -0,0 +1,148 @@
+
+		// check if the user has already accepted/read the privacy policy.
+		if ( isPolicyAlreadyAccepted === true && ! config.isEnabled( 'privacy-policy/test' ) ) {
+			return <QueryPrivacyPolicy />;
```
Doesn't the banner always needs a query on mount? What happens if they accept a policy, and then we push out another update?

retrofox replied:
Well ... the period of a new privacy policy could take time. Let's say six months. And the period used to show the banner is one month approx, it's defined in the endpoint response body, and it could be different for each version though.
But regarding the implementation, I think you're right. Maybe we should query when the component is mounted.

gwwar commented on client/blocks/privacy-policy-banner/index.jsx:
```diff
@@ -0,0 +1,148 @@
+		} );
+
+		return (
+			<div>
```
Missing the classname here.

retrofox replied:
it's weird because we don't have an eslint rule for this. It seems to be acceptable.

retrofox replied:
I had to add a css just to set up the styles so anyway it's fixed. 👍

gwwar commented on client/blocks/privacy-policy-banner/index.jsx:
```diff
@@ -0,0 +1,148 @@
+
+		return (
+			<div>
+				<QueryPrivacyPolicy />
```
I'd recommend updating the if checks above to render null. If it passes through those always render `<QueryPrivacyPolicy />` but add boolean flags next to Banner and PrivacyPolicyDialog as needed.

```jsx
{ showPrivacyPolicyBanner && <Banner
```

gwwar commented on package.json:
```diff
@@ -83,6 +83,7 @@
     "hash.js": "1.1.3",
     "he": "0.5.0",
     "html-loader": "0.4.0",
+    "html-to-react": "1.3.1",
```
Do we really need this one? Was there some problem inserting content? We generate the privacy policy so it should be safe.

Note that if we do need to add a new package, the shrinkwrap must also be updated.

retrofox replied:
> Was there some problem inserting content? We generate the privacy policy so it should be safe.

Do you mean using the dangerouslySetInnerHTML attribute? It's ok to me. The thing is that we have an eslint rule for this, but we could avoid it adding a comment at the top of the file.

Are you agree with this? Because I am :-p

retrofox replied:
added here 76c93dc

gwwar replied:
> Do you mean using the dangerouslySetInnerHTML attribute?

Yes, it's ok for us to do this with trusted content.

gwwar commented:
Pretty close @retrofox, only blocker here I think is if we need `html-to-react`. I think we can get away without using the dependency, but do let us know if you had a specific use case in mind.

retrofox commented:
@gwwar I've updated this patch according to the comments. I've replaced the parser by using the dangerouslySetInnerHTML attribute. Disabled the eslint rule as well.
Latest commits will be rebased once the PR is accepted (I hope).

retrofox commented:
it's already rebased. :-D

gwwar commented:
Do we need a max-width rule for the banner?

<img width="1142" alt="screen shot 2017-11-08 at 10 32 36 am" src="https://user-images.githubusercontent.com/1270189/32567475-3ee298d2-c470-11e7-8060-8fd72688446e.png">

gwwar commented:
Do we need a max-width rule for the banner?

<img width="1142" alt="screen shot 2017-11-08 at 10 32 36 am" src="https://user-images.githubusercontent.com/1270189/32567475-3ee298d2-c470-11e7-8060-8fd72688446e.png">

retrofox commented:
Thank you very much for your review @gwwar.

retrofox commented:
> Do we need a max-width rule for the banner?

Good catch.

<img src="https://user-images.githubusercontent.com/77539/32602692-60e00ece-c547-11e7-9d0a-071c25388de7.png" width="600px" />

<img src="https://user-images.githubusercontent.com/77539/32602697-65ca0d2c-c547-11e7-8cab-da6e07addf1b.png" width="600px" />

<img src="https://user-images.githubusercontent.com/77539/32602719-734cb670-c547-11e7-913d-221aa362e5b0.png" width="600px" />
